pkg/user/registry/useridentitymapping/rest.go

```diff
@@ -4,42 +4,360 @@ import (
+	oldmapping := old.(*api.UserIdentityMapping)
+	return validation.ValidateUserIdentityMappingUpdate(mapping, oldmapping)
+}
+
+// Get returns the mapping for an identity
+func (s *REST) Get(ctx kapi.Context, name string) (runtime.Object, error) {
+	_, _, _, _, mapping, err := s.getRelatedObjects(ctx, name)
+	return mapping, err
 }
 
-// Update will create or update a UserIdentityMapping
+// Create associates a user and identity if they both exist, and the identity is not already mapped to a user
+func (s *REST) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, error) {
+	mapping := obj.(*api.UserIdentityMapping)
+
+	if err := rest.BeforeCreate(Strategy, ctx, obj); err != nil {
+		return nil, err
+	}
+
+	identity, identityErr, _, _, _, existingMappingErr := s.getRelatedObjects(ctx, mapping.Name)
+
+	// Fail if a mapping for this identity already exists
+	if existingMappingErr == nil {
+		return nil, kerrs.NewAlreadyExists("UserIdentityMapping", mapping.Name)
+	}
+	if !kerrs.IsNotFound(existingMappingErr) {
+		return nil, existingMappingErr
+	}
+
+	// Ensure identity is valid
+	if kerrs.IsNotFound(identityErr) {
+		errs := kerrs.ValidationErrorList([]error{
+			kerrs.NewFieldInvalid("identity.name", mapping.Identity.Name, "referenced identity does not exist"),
+		})
+		return nil, kerrs.NewInvalid("UserIdentityMapping", mapping.Name, errs)
+	}
+	if identityErr != nil {
+		return nil, identityErr
+	}
+
+	// Ensure user is valid
+	user, err := s.userRegistry.GetUser(ctx, mapping.User.Name)
+	if kerrs.IsNotFound(err) {
+		errs := kerrs.ValidationErrorList([]error{
+			kerrs.NewFieldInvalid("user.name", mapping.User.Name, "referenced user does not exist"),
+		})
+		return nil, kerrs.NewInvalid("UserIdentityMapping", mapping.Name, errs)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// Update user to point to identity first. If this fails, the Create call is re-entrant.
+	if addIdentityToUser(identity, user) {
+		if _, err := s.userRegistry.UpdateUser(ctx, user); err != nil {
+			return nil, err
+		}
+	}
+
+	// Update identity to point to user last. If this fails, the Create call is re-entrant.
+	if setIdentityUser(identity, user) {
+		if updatedIdentity, err := s.identityRegistry.UpdateIdentity(ctx, identity); err != nil {
+			return nil, err
+		} else {
+			identity = updatedIdentity
+		}
+	}
+
+	return mappingFor(user, identity), nil
+}
+
+// Update associates an identity with a user.
+// Both the identity and user must already exist.
+// If the identity is associated with another user already, it is disassociated.
 func (s *REST) Update(ctx kapi.Context, obj runtime.Object) (runtime.Object, bool, error) {
-	mapping, ok := obj.(*api.UserIdentityMapping)
-	if !ok {
-		return nil, false, fmt.Errorf("not a user identity mapping: %#v", obj)
+	mapping := obj.(*api.UserIdentityMapping)
+
+	Strategy.ResetBeforeCreate(mapping)
+	identity, _, oldUser, _, oldMapping, oldMappingErr := s.getRelatedObjects(ctx, mapping.Name)
```

Comment:
Validate that the names match up correctly between this and identity.  Otherwise this method may not behave how a caller expects.